### ISSUE TITLE
[Feature Request] [lxqt-panel] Add expanding taskbar

### ISSUE THREAD

**Issue by ryanrms** [lxqt/lxqt#1553]
If you need an example of this behaviour, check ouf the Xfce4 panel.  There's an option for a fixed size that grows as items are added.  Useful for Deskbar (vertical) mode.
##### Expected Behavior

Panel should have an option to grow the panel as items are added, along with a set minimum size

##### Current Behavior

![feature2](https://user-images.githubusercontent.com/2936638/44931924-924eda00-ad20-11e8-8da0-8a75982aae3e.png)
Panel squishes taskbar items to fit in current space
##### Possible Solution

![feature1](https://user-images.githubusercontent.com/2936638/44931933-9c70d880-ad20-11e8-800b-2da9ce0a52c7.png)
Taskbar grows to accomodate new items instead of squishing them into the same area (here I've changed the panel length manually)

##### Steps to Reproduce (for bugs)

1. Set taskbar in vertical mode
2. Shorten it to only take up part of the screen and set taskbar to horizontal mode
3. Open a few application windows
4. Observe how they are squished together

##### Context

Just a nice feature from Xfce that I'd love to see in LXQT
##### System Information

* Distribution & Version: Debian testing "buster" with experimental packages
* Kernel: Linux azusa 4.17.0-3-amd64 #1 SMP Debian 4.17.17-1 (2018-08-18) x86_64 GNU/Linux
* Qt Version: 5.11.1+dfsg-6
* liblxqt Version: Version: 0.13.0-1
* Package version: 
Package: lxqt-panel
Version: 0.13.1~907-gd03d1f05-1

**Comment by ryanrms** [comment]
I should note that this isn't just useful for the taskbar.  It's also useful for the systemtray, to expand as items are added.  So if it could be added to the panel as a whole that would be amazing.  Thanks for the great DE!

**Comment by tsujan** [comment]
Yes, this is a bug but expanding isn't a solution because, usually, there's no space for that.

A solution can be making the task-bar *scrollable* when there are too many items in it. What "too many" means depends on whether the items are rotated, the minimum number of non-elided letters allowed, etc.

**Comment by tsujan** [comment]
> It's also useful for the systemtray

The systray should have an option to hide/un-hide items, as in KDE -- a feature requested for a long time but not implemented.

**Comment by wesbluemarine** [comment]
I would really like a scrollable task-bar, my actual setup is a vertical panel with icon-only launcher + icon-only taskbar, but It's easy to have too many items.
   
![qtstep_by_wesbluemarine-dclig9i 1](https://user-images.githubusercontent.com/14043560/44944378-ff707700-add5-11e8-85c9-44d07ccad10f.png)

**Comment by tsujan** [comment]
IMO, this is a must-have, although its implementation may not be easy (the same holds for the systray, but it might be easier). Either I'm missing something or we haven't had a way of dealing with too many items on task-bars for a long time.

If a window is activated, the scrollable task-bar should scroll to its corresponding item. The task-bar should be scrolled by pressing two arrow buttons on its ends. Scrolling with mouse wheel is needed too and its conflict with the current function of the mouse wheel should be resolved. When items have text, we should decide how much they can be elided before scrolling happens (perhaps when nothing other than "..." remains). And so on.

**Comment by antis81** [comment]
Since this was brought up on the TG group I think it may be an intereresting discussion to empower (or even switch between) the size & positioning system(s) for the panel. However the topic is outlined very vaguely in this issue and the implementation is quite "feature complete" in the sense that it is fully functional.

When looking at the current implementation though this whish actually would introduce a completely new mode/behavior, which contradicts the "multiple lines" concept (-> allows for a lot of items).

The first thing that comes to mind when dealing with such issues is "Why not use/create property bindings?". Currently the values are rather independent (e.g. "panel size" vs "icon size"). Depending on the vertical/horizontal orientation "size" can become "width" or "height". Also instead of a fixed "icon size" an "icon margin" would already be more flexible etc.

For the "elide" case this could also be configurable… (@tsujan)

Also @tsujan: Do you think it is realistic to outline a more complete version of the concept and offer this to be implemented - maybe in parts by different people?

**Comment by tsujan** [comment]
> Do you think it is realistic to outline a more complete version of the concept...

It's good to start with general ideas, but the more specific a feature request is, the more probably it might be implemented (by LXQt devs or contributors).

IMHO, something like PCManFM-Qt's path-bar (with path buttons) could come handy.

**Comment by antis81** [comment]
LXQt's panel could have a better (like "common") sizing/positioning concept. Not much new here… What is challenging though is the vast diversity in (virtual) screen dimensions and what users "desire".

As you might have seen I looked through the implementation a little… Technically it is not too difficult (part of this exists): We can make use of the (existing) anchor points (border and margin) for panel plugins/items/icons/text/… This alone would already provide a quite visible improvement.

To emphasize this a little better a "functional mockup" can help a lot to not get lost in details (chances are something like this exists somewhere…??).

Not sure how feature complete other implementations are (DesQ maybe??), but the "reference" should be in a better shape…
